feat: AI workflow builder front-end (no-changelog)

OlegIvaniv opened the pull request:
## Summary

This PR implements front-end for a new AI Builder feature in n8n. The implementation maintains the existing AI Assistant while adding the new Builder functionality.

## Design Library Changes

- **Chat Component Refactoring**: Refactored `AskAssistantChat` to improve component architecture through composition
  - Split monolithic chat component into smaller message-type specific components
  - Created a cleaner component hierarchy with `BaseMessage` as the foundation
  - Added specialized message rendering components for different content types

- **New Message Types**: Added support for workflow-specific message types:
  - `WorkflowStepsMessage`: Shows workflow generation steps
  - `WorkflowNodesMessage`: Displays list of nodes to be used
  - `ComposedNodesMessage`: Shows configured node details
  - `WorkflowGeneratedMessage`: Indicates workflow completion
  - `RateWorkflowMessage`: Provides a feedback mechanism with thumbs up/down

- **Component API Improvements**:
  - Added slots for customization (`header`, `placeholder`, `inputPlaceholder`)
  - Added type definitions for new chat messages
  - Added new events for feedback handling (`thumbsUp`, `thumbsDown`, `submitFeedback`)

- **UI Enhancements**:
  - Updated `AskAssistantButton` to support both assistant and builder modes
  - Added styling for new message types and feedback components

## Editor-UI Changes

- **AI Builder Store**: Implemented a dedicated store for the AI Builder functionality
  - Implemented workflow prompt tracking
  - Added telemetry for workflow generation events
  - Added experiment flag support for feature rollout

- **AI Workflow Generation**: Added functionality to generate and insert workflows
  - Implemented API interaction with backend builder endpoint
  - Added support for workflow JSON processing and insertion into canvas
  - Added generated stickies positioning logic for documentation nodes

- **UI Components**:
  - Added new `AssistantsHub` component to manage multiple assistants
  - Created `HubSwitcher` for toggling between AI Assistant and Builder
  - Implemented `CanvasNodeAIPrompt` for workflow prompt input on canvas
  - Updated floating button to support both assistant modes

- **Integration with Canvas**:
  - Added support for AI prompt node in canvas
  - Implemented workflow import functionality
  - Connected workflow generation with canvas events
  - Added automatic layout optimization for generated workflows

The implementation uses a feature flag to control rollout through PostHog experimentation.

https://github.com/user-attachments/assets/9a89556d-bbfe-45de-8f1d-cb1db98e3ddc

## Related Linear tickets, Github issues, and Community forum posts

<!--
Include links to **Linear ticket** or Github issue or Community forum post.
Important in order to close *automatically* and provide context to reviewers.
-->
<!-- Use "closes #<issue-number>", "fixes #<issue-number>", or "resolves #<issue-number>" to automatically close issues when the PR is merged. -->


## Review / Merge checklist

- [ ] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [ ] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

OlegIvaniv committed:
AI Builder implementation (squashed)

OlegIvaniv committed:
Cleanup

OlegIvaniv committed:
Merge branch 'master' into ai-680-spike-add-node-to-canvas-wconfiguration

OlegIvaniv committed:
Merge branch 'master' into ai-680-spike-add-node-to-canvas-wconfiguration

OlegIvaniv committed:
Post-merge fixes

OlegIvaniv committed:
Insert workflow JSON and tidy up, improve loading states

OlegIvaniv committed:
Merge branch 'master' into ai-680-spike-add-node-to-canvas-wconfiguration

OlegIvaniv committed:
Merge branch 'master' into ai-680-spike-add-node-to-canvas-wconfiguration

OlegIvaniv committed:
Linting fix

OlegIvaniv committed:
Merge branch 'master' into ai-680-spike-add-node-to-canvas-wconfiguration

OlegIvaniv committed:
Add canvas ai prompt. tracking chat message and fix content output limits

OlegIvaniv committed:
Merge branch 'master' into ai-680-spike-add-node-to-canvas-wconfiguration

OlegIvaniv committed:
Add AIPrompt if experiment is active, delete it on generation

OlegIvaniv committed:
Add input placeholder & use o3 mini for initial steps

OlegIvaniv committed:
refactor: Create a new builder.store and use them based on the feature flag

OlegIvaniv committed:
Use locale in AI builder

gandreini committed:
Fixed missing new font (InterVariable) in storybook.

gandreini committed:
Prompt textarea.

gandreini committed:
Error message style improved.

gandreini committed:
Sidebar messaging style tweaks.

OlegIvaniv committed:
Map node types to names

OlegIvaniv committed:
Fix builder chat offset

OlegIvaniv committed:
Add back fallback nodes if empty

OlegIvaniv committed:
Disable prompt if in progress

OlegIvaniv committed:
Fix-up stickies position

OlegIvaniv committed:
Always show feedback

OlegIvaniv committed:
Disable prompt of streaming

OlegIvaniv committed:
Improve prompts and use gpt41mini

OlegIvaniv committed:
Use o3 mini for steps

OlegIvaniv committed:
Sync state to chat when generating workflows from canvas

gandreini committed:
Various copy updates

gandreini committed:
Removed the nodeType from the Configuring nodes step

gandreini committed:
Replaced thumb button with buttons with text

gandreini committed:
Improved Start again button and empty state

OlegIvaniv committed:
Merge branch 'master' into ai-859-t2wf-p0-design-improvements

OlegIvaniv committed:
Various design fixes

OlegIvaniv committed:
Refactor model loading

OlegIvaniv committed:
Implement assistant hub for switching builder/chat

OlegIvaniv committed:
Use locale in hub switcher

OlegIvaniv committed:
Refactor graph to use chains and pass-through correct headers

OlegIvaniv committed:
Clean up packages

OlegIvaniv committed:
Extract to ai-builder to a standalone package

OlegIvaniv committed:
Cleanup & fix tracking

OlegIvaniv committed:
Fix type errors

OlegIvaniv committed:
Merge branch 'master' into ai-876-text-to-workflow-telemetry

OlegIvaniv committed:
Cleanup

OlegIvaniv committed:
Revert CLI package changes

OlegIvaniv committed:
Self-review

OlegIvaniv committed:
Typefix

OlegIvaniv committed:
Lint fix

OlegIvaniv committed:
Update snapshots

OlegIvaniv committed:
Add basic AskAssistantBuild tests

OlegIvaniv committed:
Cleanup unused message type

OlegIvaniv committed:
Merge branch 'master' into ai-876-text-to-workflow-telemetry

OlegIvaniv committed:
Merge branch 'master' into ai-876-text-to-workflow-telemetry

OlegIvaniv committed:
Revert CLI change

OlegIvaniv committed:
Fix lockfile

OlegIvaniv committed:
Use localized strings for workflow builder messages

OlegIvaniv committed:
Self-review

MiloradFilipovic commented on packages/frontend/@n8n/design-system/src/components/AskAssistantChat/messages/TextMessage.vue:
```diff
@@ -0,0 +1,202 @@
+defineProps<Props>();
+const { t } = useI18n();
+
+const md = new Markdown({
```
These markdown configurations are repeated a few times in message components. Can we centralize them?

MiloradFilipovic commented on packages/frontend/@n8n/design-system/src/components/AskAssistantChat/messages/workflow/WorkflowNodesMessage.vue:
```diff
@@ -0,0 +1,48 @@
+	gap: var(--spacing-3xs);
+	list-style-position: outside;
+	margin: 0;
+	padding: 0 0 0 var(--spacing-s);
```
If the intention for this is not to explicitly reset other paddings, can we just use `padding-left: var(--spacing-s)` here?

MiloradFilipovic commented on packages/frontend/editor-ui/src/stores/builder.store.ts:
```diff
@@ -0,0 +1,361 @@
+	});
+
+	// No need to track unread messages in the AI Builder
+	const unreadCount = computed(() => 0);
```
Nothing serious, but it would be nice if we can show popup when response is received while sidebar is closed (especially if something went wrong during generation and you don't get a workflow in the canvas):
![image](https://github.com/user-attachments/assets/38f18e0a-9b8c-4a59-9661-6ea0b87dc4cb)

MiloradFilipovic commented on packages/frontend/editor-ui/src/stores/builder.store.ts:
```diff
@@ -0,0 +1,361 @@
+				width: window.innerWidth,
+			};
+			// Always reset the chat when closed
+			// resetBuilderChat();
```
Should this be deleted?

MiloradFilipovic reviewed (changes requested):
Looking really nice! 🙌🏻

I left a few comments in the code and also here are some of the UI/UX things that i noticed:
- Assistant/Builder sidebars each save their own width so width changes as you switch modes. I would expect those to be synced
- Since builder now starts up automatically when clicking assistant button on the canvas, some existing assistant users may not realized they are not chatting with the assistant anymore when they just open up the canvas and start typing
- Canvas prompt text area can be moved around like any other node, I guess that is not intentional? Also area next to it (that contains word `or` has pointer cursor even though it's not clickable)
- Assistant hub does not switch to builder mode when prompt is submitted from the canvas and assistant tab is active
- I am always getting a workflow, even when I enter gibberish:
![image](https://github.com/user-attachments/assets/7dbf2e22-0254-42fb-91a1-ec616de547c7)
- Overflow issues in error message (I get this when i unset proxy api key environment variables):
![image](https://github.com/user-attachments/assets/31b112ed-d3b9-4be1-a66d-d9f764c1e9f8)

OlegIvaniv committed:
Use the new auth handler from AI SDK Client

OlegIvaniv committed:
Apply review suggestions

OlegIvaniv committed:
Add prompt validation

OlegIvaniv committed:
Imporve planner prompt

OlegIvaniv committed:
Add back eslint-expect error

OlegIvaniv committed:
Fix error message word break

OlegIvaniv committed:
Imporve assistant/builder active state

OlegIvaniv committed:
Merge branch 'master' into ai-876-text-to-workflow-telemetry

OlegIvaniv committed:
Keep both stores width synced

OlegIvaniv committed:
Remove pointer cursor from "or"

OlegIvaniv commented:
Thanks for thorough review, @MiloradFilipovic. I applied your suggestions:

> Assistant/Builder sidebars each save their own width so width changes as you switch modes. I would expect those to be synced

Both sidebars width is updated when resizing
> Since builder now starts up automatically when clicking assistant button on the canvas, some existing assistant users may not realized they are not chatting with the assistant anymore when they just open up the canvas and start typing

Changed it so that the AI Assistant opens by default
> Canvas prompt text area can be moved around like any other node, I guess that is not intentional? Also area next to it (that contains word `or` has pointer cursor even though it's not clickable)

The prompt area behaves exactly like "Add first step.." pseudo>node which is also draggable, we know about this UX issue but decided to live with it for MVP.
> Assistant hub does not switch to builder mode when prompt is submitted from the canvas and assistant tab is active

Fixed
> I am always getting a workflow, even when I enter gibberish:

Added moderation chain + error if the prompt is not describing a workflow
> Overflow issues in error message (I get this when i unset proxy api key environment variables):

Fixed error word-breaking


Can you have another look please?

OlegIvaniv committed:
Update snapshots

OlegIvaniv committed:
Update assistant chat min-width

OlegIvaniv committed:
Fix chat imports

MiloradFilipovic reviewed (dismissed):
Looking good, thanks for addressing the feedback 🚀

OlegIvaniv committed:
Fix tests & only show prompt if assistant is enabled

OlegIvaniv committed:
Fix default node position to always be in the middle

OlegIvaniv committed:
Update `@n8n_io/ai-assistant-sdk` version

OlegIvaniv committed:
Merge branch 'master' into ai-876-text-to-workflow-telemetry

janober commented:
Got released with [`n8n@1.91.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.91.0)
